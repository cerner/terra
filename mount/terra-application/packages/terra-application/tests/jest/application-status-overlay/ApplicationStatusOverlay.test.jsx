import React from 'react';
import uuidv4 from 'uuid/v4';

import ApplicationStatusOverlay from '../../../src/application-status-overlay/ApplicationStatusOverlay';
import ApplicationStatusOverlayContext from '../../../src/application-status-overlay/ApplicationStatusOverlayContext';

jest.mock('uuid/v4');

describe('ApplicationStatusOverlay', () => {
  let reactUseContext;
  let statusOverlayContextValue;

  beforeAll(() => {
    uuidv4.mockReturnValue('test-id');

    /**
     * Until Enzyme is updated with full support for hooks, we need to
     * mock out the useContext implementation.
     */
    reactUseContext = React.useContext;
    React.useContext = (contextValue) => {
      if (contextValue === ApplicationStatusOverlayContext) {
        return statusOverlayContextValue;
      }
      return reactUseContext(contextValue);
    };
  });

  afterAll(() => {
    React.useContext = reactUseContext;
  });

  beforeEach(() => {
    statusOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };
  });

  it('should render status view without any data', () => {
    const wrapper = mount(
      <ApplicationStatusOverlay />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should render status view with the specified data', () => {
    const StatusViewButtons = [
      {
        text: 'Action 1',
        key: 1,
      }, {
        text: 'Action 2',
        key: 2,
      },
    ];

    const wrapper = mount(
      <ApplicationStatusOverlay
        buttonAttrs={StatusViewButtons}
        message="Status View with all props specified"
        variant="no-data"
      />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should redisplay status view with new props', () => {
    const wrapper = mount(
      <ApplicationStatusOverlay message="Error status view" variant="error" />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.setProps({ message: 'No data status view', variant: 'no-data' });

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(2);
    expect(statusOverlayContextValue.show.mock.calls[1][0]).toBe('test-id');
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();

    expect(statusOverlayContextValue.show.mock.calls.length).toBe(2);
    expect(statusOverlayContextValue.hide.mock.calls.length).toBe(1);
    expect(statusOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should honor buttonAttrs prop', () => {
    const StatusViewButtons1 = [
      {
        text: 'Action 1',
        key: 1,
      }, {
        text: 'Action 2',
        key: 2,
      },
    ];

    const StatusViewButtons2 = [
      {
        text: 'Action 3',
        key: 3,
      },
    ];

    const wrapper = mount(
      <ApplicationStatusOverlay buttonAttrs={StatusViewButtons1} />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ buttonAttrs: StatusViewButtons2 });
    expect(wrapper).toMatchSnapshot();
  });

  it('should honor message prop', () => {
    const wrapper = mount(
      <ApplicationStatusOverlay message="First message" />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ message: 'Second message' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should honor variant prop', () => {
    const wrapper = mount(
      <ApplicationStatusOverlay variant="no-data" />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'no-matching-results' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'not-authorized' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'error' });
    expect(wrapper).toMatchSnapshot();
  });
});
