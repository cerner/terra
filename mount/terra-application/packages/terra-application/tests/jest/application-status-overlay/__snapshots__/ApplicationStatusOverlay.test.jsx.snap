// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationStatusOverlay should honor buttonAttrs prop 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={
    Array [
      Object {
        "key": 1,
        "text": "Action 1",
      },
      Object {
        "key": 2,
        "text": "Action 2",
      },
    ]
  }
/>
`;

exports[`ApplicationStatusOverlay should honor buttonAttrs prop 2`] = `
<ApplicationStatusOverlay
  buttonAttrs={
    Array [
      Object {
        "key": 3,
        "text": "Action 3",
      },
    ]
  }
/>
`;

exports[`ApplicationStatusOverlay should honor message prop 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  message="First message"
/>
`;

exports[`ApplicationStatusOverlay should honor message prop 2`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  message="Second message"
/>
`;

exports[`ApplicationStatusOverlay should honor variant prop 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  variant="no-data"
/>
`;

exports[`ApplicationStatusOverlay should honor variant prop 2`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  variant="no-matching-results"
/>
`;

exports[`ApplicationStatusOverlay should honor variant prop 3`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  variant="not-authorized"
/>
`;

exports[`ApplicationStatusOverlay should honor variant prop 4`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  variant="error"
/>
`;

exports[`ApplicationStatusOverlay should redisplay status view with new props 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
  message="Error status view"
  variant="error"
/>
`;

exports[`ApplicationStatusOverlay should render status view with the specified data 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={
    Array [
      Object {
        "key": 1,
        "text": "Action 1",
      },
      Object {
        "key": 2,
        "text": "Action 2",
      },
    ]
  }
  message="Status View with all props specified"
  variant="no-data"
/>
`;

exports[`ApplicationStatusOverlay should render status view without any data 1`] = `
<ApplicationStatusOverlay
  buttonAttrs={Array []}
/>
`;
